Bluetooth: controller: fix post DLE/PHY update event length

kruithofa opened the pull request:
Fix the controller implementation to perform connection event length reservation based on the completed Data Length Update and/or PHY Update Procedure.

Note: some optimisation is possible with regards to the length allocated, currently the maximum rx and tx time are used, which leaves a lot of unused airtime in case that the 2M phy is used. If necessary this will be fixed in a separate PR
fixes #21995
fixes #52393
Fixes #53049

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

Commit:
Bluetooth: controller: fix post DLE/PHY update event length

Fix the controller implementation to perform connection
event length reservation based on the completed Data Length
Update and/or PHY Update Procedure.

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

kruithofa commented on subsys/bluetooth/controller/ll_sw/lll_conn.h:
```diff
@@ -134,6 +136,11 @@ struct lll_conn {
 
+#if defined(CONFIG_BT_CTLR_DATA_LENGTH) || defined(CONFIG_BT_CTLR_PHY)
+	uint8_t evt_len_upd:1;
+	uint8_t evt_len_adv:1;
```
Variables renamed

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1841,9 +1846,57 @@ void ull_conn_done(struct node_rx_event_done *done)
+#if defined(CONFIG_BT_CTLR_DATA_LENGTH)
+		tx_time = lll->dle.eff.max_tx_time;
+		if (lll->evt_len_adv) {
+			rx_time = lll->dle.remote.max_rx_time;
```
Updated

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1841,9 +1846,57 @@ void ull_conn_done(struct node_rx_event_done *done)
+		lll->evt_len_upd = 0;
+#if defined(CONFIG_BT_CTLR_PHY)
+		ready_delay = (lll->role) ?
+			lll_radio_rx_ready_delay_get(lll->phy_rx, 1) :
```
done

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1841,9 +1846,57 @@ void ull_conn_done(struct node_rx_event_done *done)
+		tx_time = MAX(PDU_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, 0, 0),
+			      PDU_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, 0, lll->phy_tx));
+		rx_time = MAX(PDU_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, 0, 0),
+			      PDU_MAX_US(PDU_DC_PAYLOAD_SIZE_MIN, 0, lll->phy_rx));
```
done

kruithofa commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1841,9 +1846,57 @@ void ull_conn_done(struct node_rx_event_done *done)
+			rx_time = PDU_MAX_US(lll->dle.eff.max_rx_octets, 0, 0);
+		}
+#endif /* CONFIG_BT_CTLR_PHY */
+		ticks_slot = HAL_TICKER_US_TO_TICKS(OFFS_UNIT_300_US + ready_delay + EVENT_IFS_US + rx_time + tx_time + 4);
```
done

erbr-ot commented:
A thought - couldn't we know (ie remember) the sum of current ready_delay and tx/rx time contributions, so that we only need to calc the difference (to get the slot_plus/minus) to perform the ticker update? Instead of doing the 'full monty'?

kruithofa commented:
> A thought - couldn't we know (ie remember) the sum of current ready_delay and tx/rx time contributions, so that we only need to calc the difference (to get the slot_plus/minus) to perform the ticker update? Instead of doing the 'full monty'?

We could, but it is a trade-off between memory usage and processor time. These calculations are only done occasionally, so no big effect on processing time

carlescufi commented:
@cvinayak please take another look

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1874,8 +1929,8 @@ void ull_conn_done(struct node_rx_event_done *done)
-					      ticks_drift_plus,
-					      ticks_drift_minus, 0, 0,
+					      ticks_drift_plus, ticks_drift_minus,
+					      ticks_slot_plus, ticks_slot_minus,
```
In my perspective, updating the slot reservation based on DLE `max_octets` is incorrect for this controller. The ticker shall only reserve the time needed for basic LL transactions, i.e. for a PDU size of 27 bytes in each direction.
Using `max_octets` will cause too heavy reservations for ACL connections, causing substantial scheduling collisions.

Commit:
Bluetooth: controller: Kconfig for event length update

Since the event length update is not necessarily an improvement
in all situations a Kconfig option is added so that it can
be disabled for the users that do not need it.

Signed-off-by: Andries Kruithof <andries.kruithof@nordicsemi.no>

ahmedmoheb-nordic commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -1865,8 +1871,57 @@ void ull_conn_done(struct node_rx_event_done *done)
+			ticks_slot_minus = conn->ull.ticks_slot - ticks_slot;
+		}
+		conn->ull.ticks_slot = ticks_slot;
+		printk("Ticks_slot after: %d\n", ticks_slot);
```
Is that print message intended to be merged?

kruithofa replied:
good catch

mtpr-ot commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -459,6 +459,14 @@ config BT_CTLR_CENTRAL_SPACING
+config BT_CTLR_CONN_EVENT_LEN_UPDATE
+	bool "Update event length reservation after PHY or DLE update"
+	depends on !BT_LL_SW_LLCP_LEGACY && (BT_CTLR_DATA_LENGTH || BT_CTLR_PHY)
+	default n
```
I think the recommendation is to not set `default n`, as it is in fact default. Also I'm wondering if the name should reflect that we're talking about slot reservation?

kruithofa replied:
You're probably right, a few issues popped up recently that will be fixed by this PR
As for the name, I am open to suggestions, or how about BT_CTLR_SLOT_RESERVATION_UPDATE

carlescufi commented:
@mtpr-ot @cvinayak please review

mtpr-ot reviewed (commented):
LGTM
